kernel/condvar.c

```diff
@@ -0,0 +1,62 @@
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <toolchain.h>
+#include <ksched.h>
+#include <wait_q.h>
+
+static struct k_spinlock lock;
+
+int k_condvar_init(struct k_condvar *condvar)
+{
+	z_waitq_init(&condvar->wait_q);
+	z_object_init(condvar);
+	return 0;
+}
+
+int k_condvar_signal(struct k_condvar *condvar)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	struct k_thread *thread = z_unpend_first_thread(&condvar->wait_q);
+
+	if (thread != NULL) {
+		arch_thread_return_value_set(thread, 0);
+		z_ready_thread(thread);
+	}
+	z_reschedule(&lock, key);
+	return 0;
+}
+
+int k_condvar_broadcast(struct k_condvar *condvar)
+{
+	struct k_thread *pending_thread;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+	/* wake up any threads that are waiting to write */
+	while ((pending_thread = z_unpend_first_thread(&condvar->wait_q)) !=
+	       NULL) {
+		arch_thread_return_value_set(pending_thread, -ENOMSG);
+		z_ready_thread(pending_thread);
+	}
+
+	z_reschedule(&lock, key);
+
+	return 0;
+}
+
+int k_condvar_wait(struct k_condvar *condvar, struct k_mutex *mutex,
+		   k_timeout_t timeout)
+{
+	int ret;
+	k_spinlock_key_t key;
+
+	k_mutex_unlock(mutex);
+	key = k_spin_lock(&lock);
```

Comment:
The standard semantics of condition variables requires that the unlock, queue, and wait be done atomically.  This code leaves a gap where the signal could be sent before the thread is queued to receive it, causing a failure.

You'd need to move the spin lock before the mutex to avoid that.

Reply:
yep, missed that.